// @parcel/transformer-typescript-tsc/lib/TSCTransformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _plugin = require("@parcel/plugin");

var _tsUtils = require("@parcel/ts-utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _default = new _plugin.Transformer({
  async loadConfig({
    config,
    options
  }) {
    await (0, _tsUtils.loadTSConfig)(config, options);
  },

  async transform({
    asset,
    config,
    options
  }) {
    asset.type = 'js';
    let [typescript, code] = await Promise.all([options.packageManager.require('typescript', asset.filePath), asset.getCode()]);
    let transpiled = typescript.transpileModule(code, {
      compilerOptions: _objectSpread({
        // React is the default. Users can override this by supplying their own tsconfig,
        // which many TypeScript users will already have for typechecking, etc.
        jsx: 'React'
      }, config, {
        // Always emit output
        noEmit: false,
        // Don't compile ES `import`s -- scope hoisting prefers them and they will
        // otherwise compiled to CJS via babel in the js transformer
        module: typescript.ModuleKind.ESNext
      }),
      fileName: asset.filePath // Should be relativePath?

    });
    return [{
      type: 'js',
      code: transpiled.outputText
    }];
  }

});

exports.default = _default;
